python/lsst/ts/watcher/rules/cp_verify_alarm.py

```diff
@@ -0,0 +1,305 @@
+                for test in fail_count:
+                    if test in counter:
+                        counter[test] += 1
+                    else:
+                        counter[test] = 1
+                total_counter_failed_tests[exposure] = counter
+            else:
+                continue
+
+        # If there are not exposures with tests that failed.
+        if len(total_counter_failed_tests) == 0:
+            return certify_calib, None, None
+
+        # Count the number of exposures where a given test fails
+        # in the majority of detectors.
+        failed_exposures_counter = 0
+        for exposure in total_counter_failed_tests:
+            for test in total_counter_failed_tests[exposure]:
+                if (
+                    total_counter_failed_tests[exposure][test]
+                    >= failure_threshold_exposure
+                ):
+                    failed_exposures_counter += 1
+                    # Exit the inner loop over tests: just need
+                    # the condition to be satisfied for
+                    # at least one type of test
+                    break
+
+        # For at least one type of test, if the majority of tests fail in
+        # the majority of detectors and the majority of exposures,
+        # then don't certify the calibration
+        if failed_exposures_counter >= max_number_failed_exposures:
+            certify_calib = False
+
+        # Return a dictionary with the thresholds to report
+        # them if verification fails.
+        thresholds = {
+            "MAX_FAILURES_PER_DETECTOR_PER_TEST_TYPE_THRESHOLD": max_number_failures_per_detector_per_test,
+            "MAX_FAILED_DETECTORS_THRESHOLD": max_number_failed_detectors,
+            "MAX_FAILED_TESTS_PER_EXPOSURE_THRESHOLD": failure_threshold_exposure,
+            "MAX_FAILED_EXPOSURES_THRESHOLD": max_number_failed_exposures,
+            "FINAL_NUMBER_OF_FAILED_EXPOSURES": failed_exposures_counter,
+        }
+
+        return certify_calib, total_counter_failed_tests, thresholds
+
+    def get_cp_verify_stats(self, response_verify):
+        """Get cp_verify statistics from the butler.
+
+        Parameters
+        ----------
+        response_verify : `dict`
+            OCPS call response.
+
+        Returns
+        -------
+        verify_stats : `dict`
+            Statistics from cp_verify.
+        """
+        job_id_verify = response_verify["job_id"]
+        # Loop over the entries of the 'results' list
+        # and look for the adecuate dataset type.
+        for entry in response_verify["results"]:
+            if "verifyBiasStats" in entry["uri"]:
+                verify_stats_string = "verifyBiasStats"
+                break
+            elif "verifyDarkStats" in entry["uri"]:
+                verify_stats_string = "verifyDarkStats"
+                break
+            elif "verifyFlatStats" in entry["uri"]:
+                verify_stats_string = "verifyFlatStats"
+                break
```

Comment:
Please don't keep extracting `entry["uri"]`. Do it once at the beginning.

I also suggest a regex to simplify the code:
```
import re
...
match = re.search(r"(verify(Bias|Dark|Flat)Stats)", entry["uri"])
if match is None:
    raise ...
verify_stats_str = match.groups()[0]
```

Or if you prefer to avoid regexes, consider a loop such as:
```
uri = entry["uri"]
for substr in ("verifyBiasStats", "verifyDarkStats", "verifyFlatStats"):
    if substr in uri:
        verify_stats_string = substr
        break
else:
    raise ...
```
The `else` clause in the `for` statement is a slightly obscure Python feature. You could avoid it by initializing `verify_stats_string` to None and testing for that after the loop.